### ISSUE TITLE
Selected inputs reset on connection to Server with v3.7.0

### ISSUE THREAD

**Issue by pgScorpio** [jamulussoftware/jamulus#1346]
**Describe the bug**

Selected input channels always reset to the first two channels as soon as you connect to a server.

**To Reproduce**

Using ASIO Link Pro, with more than 2 channels, as device for Jamulus.
Now in Jamulus settings select, channels lets say. LinkIn3,LinkIn4 and LinkOut3, LinkOut4.
Now Connect to a server. 
As soon as the connection is established the selected channels revert to LinkIn1,LinkIn2 and LinkOut1, LinkOut2.
(After that you can select LinkIn3,LinkIn4 and LinkOut3, LinkOut4 again and it will work as expected.)

**Expected behavior**

No changing of selected channels when connecting to a server.

**Screenshots**

**Operating system**

Windows 10

Strangely enough it reproduces consequently on my PC:
Dutch Windows 10 Pro, build 19041
AMD Athlon II X2 215, 6GB RAM, 512GB SSD

But it doesn't reproduce on my laptop:
English Windows 10 Home, build 19042
AMD A4-5000, 8GB RAM, 1TB HDD

**Version of Jamulus**

jamulus v3.7.0 downloaded via Jamulus.io (https://github.com/jamulussoftware/jamulus/releases/download/r3_7_0/jamulus_3.7.0_win.exe)

**Additional context**

A lot of choirs use ASIO Link Pro to connect the audio between Jamulus and Zoom. (See my JamZoom-ALP.pdf document in the Jamulus Choral Community on Facebook.)
As far as I've heard until now about 60% to 70% of these ALP/Jamulus installs seem to show this problem with v3.7.0
Reverting back to Jamulus 3.6.2 always solves the problem.

**Comment by ann0see** [comment]
Probably this is related to a bug fix which was done to solve a bug in ASIO4ALL. #796 

Channel mapping should only be reset if device properties change, not if you connect to a server (and I thought this would be the case). Not yet sure how we can solve this issue since the previous approach was even worse (and might have resulted in a crash of Jamulus if you change inputs/outputs in ASIO4ALL).

**Comment by ann0see** [comment]
Maybe, not totally sure, we could re-implement the removed if and check if the number of channels changed. This might solve both bugs.

**Comment by ann0see** [comment]
@pgScorpio once the compilation in my PR #1347  is finished, could you please check if the changes I made solved your issue?

**Comment by pgScorpio** [comment]
> Probably this is related to a bug fix which was done to solve a bug in ASIO4ALL. #796
very well possible....

> 
> Channel mapping should only be reset if device properties change, not if you connect to a server (and I thought this would be the case). Not yet sure how we can solve this issue since the previous approach was even worse (and might have resulted in a crash of Jamulus if you change inputs/outputs in ASIO4ALL).
I always disconnected from server before changing ASIO settings to avoid crashing of Jamulus (Which happens very often), so maybe automatically disconnecting when opening the ASIO settings would be a dirty fix.

> 
> The first difference between your pc and your laptop is that they use different versions of windows (Pro vs. Home).
Unfortunately most people that reported this problem to me used Windows Home so that doesn't add up. The only thing I noticed is that all people having this problem appear to be using a Dutch Windows version, but that doesn't make any sense to me too.

**Comment by ann0see** [comment]
Hmm. In the meantime, I made some changes to the code. Can you check the Pull request?

**Comment by pgScorpio** [comment]
> @pgScorpio once the compilation in my PR #1347 is finished, could you please check if the changes I made solved your issue?

If you can give me a link to the changed version I will be happy to test it.

**Comment by ann0see** [comment]
Ok. I just compiled it locally so you can test it: https://github.com/ann0see/jamulus/releases/download/rDebugReset_2021-03-25/Jamulus-3.7.0dev-installer-win.exe

Be aware that this might not be the behaviour of the final version, but it's worth a try.

**Comment by ann0see** [comment]
> See my JamZoom-ALP.pdf document in the Jamulus Choral Community on Facebook.

Is this file freely available? Can you add the content of it to the community knowledge base? It would be very valuable to have it there

**Comment by pgScorpio** [comment]
> Ok. I just compiled it locally so you can test it: https://github.com/ann0see/jamulus/releases/download/rDebugReset_2021-03-25/Jamulus-3.7.0dev-installer-win.exe
> 
> And https://github.com/ann0see/jamulus/releases/tag/rDebugReset_2021-03-25 (should be the final version)
> Be aware that this might not be the behaviour of the final version, but it's worth a try.

Just did a quick test and the problem is solved with this version !

**Comment by ann0see** [comment]
Great. Can you also check if that's the case from my PR? (See the Checks page and download the artifact)?

**Comment by pgScorpio** [comment]
> > See my JamZoom-ALP.pdf document in the Jamulus Choral Community on Facebook.
> 
> Is this file freely available? Can you add the content of it to the community knowledge base? It would be very valuable to have it there

As said it's available in the documents section of the "Jamulus choral community" on facebook.
And a Dutch version in the documents section of the Dutch "Tip&Tricks online repetities" group.

For now I prefer to keep them only there since it's still in development and the documents refer to my own server at home and my email address, so for now I like to keep the distribution restricted.

**Comment by ann0see** [comment]
Ok sure.

I personally don't use/want to use Facebook but am interested in it ;-). 

If one day you decide to make it or parts of it public you're welcome.

**Comment by pgScorpio** [comment]
> Great. Can you also check if that's the case from my PR? (See the Checks page and download the artifact)?

That one is OK too.

**Comment by JoeSouthern** [comment]
Hi there. To weigh in on this.

I was having the same issue - notice 'was having' ha ha  - doesn't appear any more.

Basically my inputs were getting reset. If I used the ASIO4All they didn't. Using Presonus they did. (I've never had trouble with the Presonus).

I was actually in the process of going through the code trying to see what was up with it.

I recompiled yours - with QT 5.12.4 MSVC2015 64-Bit.  I then just renamed the old exe file and added the new exe file in the Jamulus directory.
It seems fine now. Previously, sometimes It wouldn't go but a few moments before it reset - even without being connected to a server.

Here is my original post - https://github.com/jamulussoftware/jamulus/discussions/1386

Now, it appears to me though as a side note, that theses tags

 <sndcrdinlch>28</sndcrdinlch>
 <sndcrdinrch>29</sndcrdinrch>
 <sndcrdoutlch>18</sndcrdoutlch>
 <sndcrdoutrch>19</sndcrdoutrch>

which would represent the Input and Outputs - Left and Right - for my particular system, are only written to the ini file on closing of the application. Doing the save Mixer settings from the menu - produces a file that doesn't have those tags. I thought that's what that was for, but I'm not sure.

Anyway, that fixes it for me it seems. I'll post back if I run into something later, but it's been an hour or so and settings are still holding.
